### PR TITLE
Post Scheduling

### LINKED ISSUE #6413: Post Scheduling: Cron Adapter
For a full overview of the spec, please see [Post Scheduling 101](https://github.com/TryGhost/Ghost/wiki/Post-Scheduling-101) and the epic (parent) issue: https://github.com/TryGhost/Ghost/issues/6399.

----

The role of [The Cron](https://github.com/TryGhost/Ghost/wiki/Post-Scheduling-101#the-cron) is to be responsible for keeping time in a reliable way. This should be implemented as a pluggable/replaceable adapter so that different environments can drop in their own implementations based on their requirements.

## Adapter Layer

**Messages from the Scheduler**

The adapter layer needs to define the set of methods that a cron adapter must implement. These should provide a simple API for scheduling, rescheduling and unscheduling 'jobs'. The Cron can expect to be passed the following information:

- the time the job is scheduled for (datetime, or cron time)
- the API URI to call at the given time

The URI to call passed from the scheduler will be unique to a job, such that it's possible to determine which item is being [re|un]scheduled. The adapter will need to receive the details of the client id and secret it should use to authenticate its calls to the API URI. 

**Retrieving info from The Scheduler**

The Scheduler also provides a way for The Cron to retrieve information about scheduled jobs in the form of an API endpoint [see here](https://github.com/TryGhost/Ghost/wiki/Post-Scheduling-101#providing-information). This is so that it can do some initial processing when it is first started, or recover from failure. 

It may make sense for this endpoint + the client id and secret to be provided to The Cron when it is constructed. Note that Ghost API endpoints can be used via method call or HTTP.

**Defining the active adapter** 
In a very similar way to the storage adapter, we need a way to determine which adapter is in use and to pass any configuration that the adapter needs. Configuration should follow the same pattern as [storage](https://github.com/TryGhost/Ghost/wiki/Using-a-custom-storage-module#configjs).

Ideally, we'd also have a way to check that a provided adapter implements all of the required methods, or otherwise falls back to the default.

We're looking at moving towards having adapters be true npm modules. There was also a plan to refactor the storage adapter so that the base file was an npm module that would be required by any implementation of it [see example implementation](https://github.com/vdemedes/Ghost-Storage-Base). I think it's worthwhile considering this approach for the cron adapter, even though it's not as complex as the storage adapter. 

## Default Cron Implementation

The intention for the default cron is to make use of the [node-cron](https://github.com/ncb000gt/node-cron) module to handle scheduling jobs using the [hybrid strategy](https://github.com/TryGhost/Ghost/wiki/Post-Scheduling-101#strategy-e-hybrid). 

E.g. run a job at a regular interval, which asks the scheduler for items which were missed (Scheduled in the past) or are due before the job runs again. The items returned will then be scheduled to happen at the right time. The interval should be 5minutes by default, but this should ideally be configurable in config.js.

As part of an initial implementation, it would be useful to test out how node-cron works when used with 100s/1000s of jobs and if it can handle that sort of scale.

The default cron should send requests to the scheduler URIs at the correct time, and listen for the response. If it gets a 200 or a 404, we can consider the job 'done'. If the response is an error, it would be preferable to retry 3 times, at a sensible interval, before giving up.

### PR DESCRIPTION
**Pull request by kirrg001**
closes #6413

This PR contains cherry picks:
- #6737, #6759 , #6733

This PR relies on:
https://github.com/TryGhost/Ghost-Admin/pull/25
### TODO:
- [x] there are still some minor general fixes i could extract into separate PR's 
- [x] add an extra safety check into the api layer to ensure that this task should really get published at this time (and maybe some more checks)
- [x] extract config_spec fixes
- [x] change messaging for post in the future
- [x] add force flag to default scheduler
- [x] reproduce markdown
- [x] shutdown the scheduler on purpose while posts should get published, then restart
- [x] look for a solution for crash scheduler and disallow publish posts in the past because of wrong TZ configuration of sender
- [x] publish tolerance of 2minutes (?)
- [x] fix bug in publish scheduler
- [x] add x clear cache header 
  (https://github.com/TryGhost/Ghost/blob/9ec8c886900f685fa720bae3595a5175508e27f7/core/server/api/index.js#L70-L70)
- [x] one basic test with pg
- [x] big manual test with default scheduler
- [x] big manual test with pro scheduler
- [x] one final big test

### CODE REVIEW

**Comment by ErisDS** [comment]
The UI can certainly go into a separate PR btw

I am not sure, but I think we need some extra model-logic in the save function for a post.

If a post has a `status` of "scheduled" and a `published_at` date that is either in the past or ~now, and it gets saved, it should probably automatically be set to `published`?? I think we should not try to schedule a job for a post that should have been published already or in the next few seconds. 

I'd suggest that ~now be 60 seconds? E.g. If I set a post to publish in <= 60seconds, and then save it, the post should be published now, otherwise it is scheduled?

**Comment by kirrg001** [comment]
Yes UI --> Ghost Admin PR.

Right now the logic is very simple and i would keep as it is: if you wanne schedule a post with published_at < now + 5min, you will get a validationerror. Ember will show an error (and should maybe also support not picking any date under now + 5minutes. Do you agree?

(Sending scheduled + date in the past is no use case --> validation error)

**Comment by ErisDS** [comment]
I think this needs to be handled in the model rather than throwing validation errors:

The use case I'm thinking of, is someone who wants to make some last minute additions:

If we throw a validation error for a scheduled post and any date under now + 5minutes

- I scheduled my post a week ago, everything is fine
- 10 minutes before it goes live, I think of some changes I want to make
- I go to the editor and start typing
- 5 minutes later, I start getting validation errors and I can't save because my post is scheduled for in less than 5 minutes... 

If we have server side handling of this case, and set the ~now value to be +5 minutes:

- I scheduled my post a week ago, everything is fine
- 10 minutes before it goes live, I think of some changes I want to make
- I go to the editor and start typing
- 5 minutes later, my post is suddenly published when I'm just trying to save

If we have server side handling of this case and set the now value to be +60 seconds (or perhaps 30?)

- I scheduled my post a week ago, everything is fine
- 10 minutes before it goes live, I think of some changes I want to make
- I go to the editor and start typing
- I keep typing and saving, and everything is ok
- 1min or 30 seconds before the correct time, my post gets converted to published when I save.

This is still a bit weird, but 30-60 seconds I think will feel ok, whereas I think 5mins will feel wrong?

In a future version, we can add extra UI that indicates what's going on more clearly.

**Comment by kirrg001** [comment]
If a user updates the posts content, it should not affect scheduling at all. So you are right the author needs to be able to always update the content, anytime.
But I would not publish a post 60 seconds before, thats weird.

Here is my suggestion:
I would extend my condition to say "if published_at changed AND the new published_at is lower then x minutes" (maybe not 5, maybe 2). So no validation error at all (even if you update your post 1minute before).

And you can always unschedule.

**Comment by ErisDS** [comment]
> If a user updates the posts content, it should not affect scheduling at all. 

Just in case you hadn't realised, our API doesn't do partial updates (no PATCH support). Whenever a post is saved, every property is sent, so you would have the scheduling properties sent at the same time.

**Comment by kirrg001** [comment]
- you create a new post as draft
- you configure your published_at to tomorrow 11AM
- you click "schedule post" (because you can, because published_at is in the future)
- 10 minutes before it gets scheduled, you click "update post" to update your model because of a content change
- now ember sends the new post model (published_at did not change, status did not change!)
- server figures out: published_at did not change, status did not change, ok go through with saving
- scheduling is not touched at all, no rescheduling event, no error, just a content update
- post gets updated, server returns 200

**Comment by kirrg001** [comment]
Ok maybe i know now what your concern is. If the scheduler updates the post model (to published) and the ember app has an older version of the model (in case he was all he time in the admin section waiting that his post get scheduled), he would reset the model to status scheduled. In my opinion is this a general problem and therefor we might need versioning.

**Comment by kirrg001** [comment]
conclusion for V1 (because we don't support versioning):
- if ember client sends scheduled and the post is already published, we disallow and throw an error
- we reduce 5minutes to something lower and throw a validation error if ember wants to reschedule a post 1 or 2 minutes before the posts get scheduled (the user then needs to unschedule)

File: core/server/models/base/when.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+var config = require('../../config'),
```
Comment (kirrg001) [inline core/server/models/base/when.js]: The basic idea of `when` is to avoid listening on events all over the server.
I would love to create a centralized place, where we listen on all kinds of events and then delegate.
But for now, we could also move this logic into `scheduling/post-scheduling`.

File: core/client [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit 4b3bc50b04cd28f303e6981e2f027c3eef24e592
+Subproject commit 61fe83dff375a145d5272a5382aa2c72bd994830
```
Comment (kirrg001) [inline core/client]: revert this change!

File: core/server/scheduling/post-scheduling/index.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    return _private.loadClient()
+        .then(function (_client) {
+            client = _client;
+
+            return utils.createAdapter(config);
+        })
+        .then(function (_adapter) {
+            adapter = _adapter;
+
+            return _private.loadScheduledPosts();
```
Comment (kirrg001) [inline core/server/scheduling/post-scheduling/index.js]: `loadScheduledPosts` is also triggered when we run it on `Ghost(Pro)`.
This has the advantage of integrity and has the disadvantage that we maybe just reinsert the jobs.
We cannot move this piece of logic into the `SchedulingDefault`, because this adapter can be used also for newsletter scheduling. So i just kept it here and we can discuss and decide.

File: core/server/models/base/when.js (hunk 1 above)
Comment (kirrg001) [inline core/server/models/base/when.js]: i saw that david also created a similar file called `events`, would create a task and merge them together

File: core/server/models/base/when.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        return;
+    }
+
+    models.Post.findAll({filter: 'status:scheduled', context: {internal: true}})
+        .then(function (results) {
+            if (!results.length) {
+                return;
+            }
+
+            return Promise.all(results.map(function (post) {
```
Comment (kirrg001) [inline core/server/models/base/when.js]: `mapSeries`

File: core/server/data/db/connection.js [hunk 5]
```diff
@@ -47,8 +46,8 @@ function configure(dbConfig) {
     return dbConfig;
 }
 
-if (!knexInstance && dbConfig && dbConfig.client) {
-    knexInstance = knex(configure(dbConfig));
+if (!knexInstance && config.database && config.database.client) {
```
Comment (ErisDS) [inline core/server/data/db/connection.js]: This looks like a change that could be pulled out?

File: core/server/models/base/when.js (hunk 1 above)
Comment (ErisDS) [inline core/server/models/base/when.js]: I think that this change could be easily moved out into its own PR on top of master, merging with the events file? There's no reason why it can't be merged before this pr?

File: core/server/models/post.js [hunk 6]
```diff
@@ -183,7 +192,9 @@ Post = ghostBookshelf.Model.extend({
 
         ghostBookshelf.Model.prototype.saving.call(this, model, attr, options);
 
-        this.set('html', converter.makeHtml(_.toString(this.get('markdown'))));
+        if (this.get('markdown')) {
```
Comment (ErisDS) [inline core/server/models/post.js]: What is the usecase for this guard? Atm Ghost doesn't really support not sending markdown?

**Comment by ErisDS** [comment]
This PR contains a lot of changes to model logic + new API endpoints + the new scheduling adapters.

I'd love to see all of this new functionality documented in the commit & PR message so that it's easier to reason about what's intended to be happening & also for future reference.

There are a few things in here which were never spec'd or are different to the specs, so it'd be good to get more explanation around what has been changed, added, decided on etc and why. I'll add some comments with a bunch of individual questions to explain what sort of information I mean.

File: core/server/index.js [hunk 7]
```diff
@@ -108,6 +112,14 @@ function init(options) {
...
 
         return new GhostServer(parentApp);
+    }).then(function (_ghostServer) {
+        ghostServer = _ghostServer;
+
+        // scheduling can trigger api requests, that's why we initialize the module after the ghost server creation
+        // scheduling module can create x schedulers with different adapters
+        return scheduling.init(_.extend(config.scheduling, {apiUrl: config.url + config.urlFor('api')}));
+    }).then(function () {
+        return ghostServer;
```
Comment (ErisDS) [inline core/server/index.js]: This doesn't seem like the right place for this to me? 

Putting it here means that scheduling won't work for people using Ghost as an npm module, unless they also update their index.js files.

If I understand correctly the idea is that we need to initialise scheduling after Ghost is ready to start receiving requests, because the scheduler might decide to start sending API requests. However, would that not be better handled via an event listener on when the server has started?

Comment (ErisDS) [inline core/server/index.js]: Hmm... actually I think I'm wrong about the npm module thing...

File: core/server/data/db/connection.js (hunk 5 above)
Comment (kirrg001) [inline core/server/data/db/connection.js]: 👍

File: core/server/models/base/when.js (hunk 1 above)
Comment (kirrg001) [inline core/server/models/base/when.js]: 👍

File: core/server/config/index.js [hunk 8]
```diff
@@ -179,8 +192,14 @@ ConfigManager.prototype.set = function (config) {
             availableApps:    this._config.paths.availableApps || {},
             clientAssets:     path.join(corePath, '/built/assets/')
         },
+        scheduling: {
+            postScheduling: {
```
Comment (ErisDS) [inline core/server/config/index.js]: I'm not sure what the reasoning behind multiple kinds of nested scheduling adapters is? This seems unnecessary to me.

File: core/server/models/post.js (hunk 6 above)
Comment (kirrg001) [inline core/server/models/post.js]: will try to reproduce, i had a case
memo to me: always add a comment ;)

File: core/server/models/post.js [hunk 9]
```diff
@@ -157,7 +166,7 @@ Post = ghostBookshelf.Model.extend({
                 return Promise.reject(new errors.ValidationError(
                     i18n.t('errors.models.post.expectedPublishedAtInFuture')
                 ));
-            } else if (moment(publishedAt).isBefore(moment().add(5, 'minutes'))) {
+            } else if (this.hasChanged('published_at') && moment(publishedAt).isBefore(moment().add(2, 'minutes'))) {
```
Comment (ErisDS) [inline core/server/models/post.js]: This is a guard with an arbitrary number... should it be defined in config somewhere, even if it isn't overridable?

Some more CASE comments here would be handy - this one in particular is quite hard to read

File: core/server/index.js (hunk 7 above)
Comment (kirrg001) [inline core/server/index.js]: I could test installing Ghost via npm. Tell me if you still have any concerns, even when they are small.

File: core/server/config/index.js (hunk 8 above)
Comment (kirrg001) [inline core/server/config/index.js]: the idea is to group all schedule configuration in one place.
`{ newsletter: { active: 'my-adapter' }, postScheduling: { active: 'pro-adapter' } }`

File: core/server/models/post.js [hunk 10]
```diff
@@ -137,13 +137,22 @@ Post = ghostBookshelf.Model.extend({
...
             tags = [];
 
-        // both page and post can get scheduled
+        // CASE: disallow published -> scheduled
+        // @TODO: remove when we have versioning based on updated_at
+        if (newStatus !== olderStatus && newStatus === 'scheduled' && olderStatus === 'published') {
+            return Promise.reject(new errors.ValidationError(
+                i18n.t('errors.models.post.isAlreadyPublished', {key: 'status'})
+            ));
+        }
```
Comment (ErisDS) [inline core/server/models/post.js]: I think there is a missing guard that should either be here or in the scheduling api - there's currently nothing that's checking that the scheduler is doing the right thing? (Or I haven't found it yet).

What I mean is, I would expect that if the scheduler endpoint receives a request to /post/:id/ that something would check that the post is still currently set to "scheduled" and the publish_at time is either in the past or somewhere in the near future, so that the post isn't getting scheduled incorrectly.

The case I'm thinking of, is someone unscheduling, or rescheduling a post just a few seconds before the schedule time, and the scheduler not getting the event & cancelling its job in time.

File: core/server/models/post.js (hunk 9 above)
Comment (kirrg001) [inline core/server/models/post.js]: just putting it into the config will not bring any value in my opinion?(maybe for testing purposes, but i didn't need it)
i could think of returning it the ember, so both will use the same number?but no prio right now.

Comment (kirrg001) [inline core/server/models/post.js]: will add a case comment 👍

File: core/server/api/schedules.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+    // CASE: only the scheduler client is allowed to publish (hardcoded because of missing client permission system)
+    if (!options.context || !options.context.client || options.context.client !== 'ghost-scheduler') {
+        return Promise.reject(new errors.NoPermissionError(i18n.t('errors.permissions.noPermissionToAction')));
+    }
+
+    return pipeline([
+        utils.validate('posts', {opts: utils.idDefaultOptions}),
+        function (cleanOptions) {
+            return dataProvider.Post.edit({status: 'published'}, _.omit(cleanOptions, 'data'))
```
Comment (ErisDS) [inline core/server/api/schedules.js]: Between looking at this and the module on saving event, I think there is some logic missing here for handling returning 404 for posts which are not scheduled, and also updating the X-Cache-Invalidation header logic. By this I'm referring to  [this section](https://github.com/TryGhost/Ghost/wiki/Post-Scheduling-101#scheduling-api) of the original spec.

File: core/server/models/post.js (hunk 10 above)
Comment (kirrg001) [inline core/server/models/post.js]: i thought of it, but model versioning would handle this case.
We could add it until we have versioning even if am convinced that this is a very raw case and only happens in case of logic bugs in the adapter (or server crash).

File: core/server/api/schedules.js (hunk 11 above)
Comment (kirrg001) [inline core/server/api/schedules.js]: This endpoint is only called from the scheduler. Invalidate the cache is handled by it, we send a property called `clearCache`.

Comment (kirrg001) [inline core/server/api/schedules.js]: `404` should happen automatically - this should be a default server behaviour. I can see a lot of tests returning 404 in case something wasn't found.

File: core/server/config/index.js (hunk 8 above)
Comment (kirrg001) [inline core/server/config/index.js]: example case:
We offer the scheduler as a service. but you need to pay for it. because it also provides fancy statistics.
You decide letting post-scheduling run with our service and newsletter locally, because otherwise its too expensive.

File: core/server/api/schedules.js (hunk 11 above)
Comment (ErisDS) [inline core/server/api/schedules.js]: X-Cache-Invalidation headers are what should be used by the scheduler to determine the `clearCache` property?
Also there is no handling here for a 404 for this endpoint - it will only 404 if the post is not found, not if the task is not found. 

We have to model-representation of a task but we still have logic for it. The task exists if the post is set to scheduled and the published_at date is in the past or near future.

File: core/server/config/index.js (hunk 8 above)
Comment (ErisDS) [inline core/server/config/index.js]: The intention here is to have one small moving piece that handles making tasks happen at a correct time, which can be completely agnostic of what those tasks are. 

We can and will then build out many, many features on top of that piece. The configuration should not need updating each time we add a new scheduling feature and my expectation is that each blog would only ever have one scheduling adapter installed. 

All of the fancy features would live in the tools that are leveraging the adapter, not in the adapter itself, and so there would never be any need to have more than one.

File: core/server/api/schedules.js (hunk 11 above)
Comment (kirrg001) [inline core/server/api/schedules.js]: i will add the X-Cache-Invalidation. i've added a task.
the 404 issue is the same we discussed already here: 
https://github.com/TryGhost/Ghost/pull/6861/files#r66493705 - we will add a check!

File: core/server/models/post.js (hunk 10 above)
Comment (kirrg001) [inline core/server/models/post.js]: i will add a check into the api layer

File: core/server/config/index.js (hunk 8 above)
Comment (kirrg001) [inline core/server/config/index.js]: sure will remove the nested config

File: core/server/index.js (hunk 7 above)
Comment (ErisDS) [inline core/server/index.js]: It's the index.js one level higher in the route that the npm module users have to re-implement, rather than this one. So it is not a problem. I think using events might be an improvement but one we can revisit later.

File: core/server/api/schedules.js [hunk 12]
```diff
@@ -6,25 +6,39 @@ var _ = require('lodash'),
...
         function (cleanOptions) {
-            return dataProvider.Post.edit({status: 'published'}, _.omit(cleanOptions, 'data'))
+            cleanOptions.status = 'scheduled';
+
+            return apiPosts.read(cleanOptions)
                 .then(function (result) {
-                    return Promise.resolve({post: result.toJSON(cleanOptions)});
+                    post = result.posts[0];
+
+                    // CASE: now and published_at should almost be equal (1minute tolerance)
```
Comment (ErisDS) [inline core/server/api/schedules.js]: IMO - this should accept any published_at which is in the past, no matter how long ago. If the scheduler was down for 20minutes, when it comes back up it should be able to correct itself.

In terms of future posts - I'd recommend using the same value we're using elsewhere for an "almost equal"? That is currently 2 mins?

Comment (kirrg001) [inline core/server/api/schedules.js]: That is a good use case (scheduler is down), but talking about timezones, what if somebody configures the TZ not to UTC, instead to UTC-6, the post would get scheduled 6 hours too early?

@ 2mins: sure we can use the the same minute offset for both cases 👍

Comment (ErisDS) [inline core/server/api/schedules.js]: > That is a good use case (scheduler is down), but talking about timezones, what if somebody configures the TZ not to UTC, instead to UTC-6, the post would get scheduled 6 hours too early?

I think the case of a user changing their timezone is much much smaller than the case of the scheduler being down or incorrectly configured?

There's no easy way, I think, to handle both??

Comment (kirrg001) [inline core/server/api/schedules.js]: will think about and come back soon 👍

Comment (kirrg001) [inline core/server/api/schedules.js]: see https://github.com/TryGhost/Ghost/pull/6861/commits/dbfc90dee02f32c53b01077d208c1b8f701fbec8 

i've added a `force` flag in case the service detects publishing a post in the past (in case the service was down). the API is able to react on this flag. this flag can be only used for publishing in the past.

File: core/server/models/post.js (hunk 6 above)
Comment (kirrg001) [inline core/server/models/post.js]: couldn't reproduce, thats why reverted https://github.com/TryGhost/Ghost/pull/6861/commits/c8ebd04f2a1df0b9f1a2cfdb013752ccf49b5461

File: core/test/integration/api/api_schedules_spec.js [hunk 13]
```diff
@@ -179,7 +179,7 @@ describe('Schedules API', function () {
             originalCannotScheduleAPostBeforeInMinutes = config.times.cannotScheduleAPostBeforeInMinutes;
 
             // we can insert published_at less then 5minutes
-            config.times.cannotScheduleAPostBeforeInMinutes = -5;
+            config.times.cannotScheduleAPostBeforeInMinutes = -15;
```
Comment (ErisDS) [inline core/test/integration/api/api_schedules_spec.js]: Tests shouldn't modify config directly. There is a utility for setting and restoring config, see `model_posts_spec.js` for a good example.

File: core/server/translations/en.json [hunk 14]
```diff
@@ -325,7 +325,8 @@
                 "postNotFound": "Post not found."
             },
             "job": {
-                "not-found": "Job not found."
+                "not-found": "Job not found.",
```
Comment (ErisDS) [inline core/server/translations/en.json]: Just a little cleanup thing here - this key should be camelCase i.e. `notFound` please

File: core/test/integration/api/api_schedules_spec.js (hunk 13 above)
Comment (kirrg001) [inline core/test/integration/api/api_schedules_spec.js]: i've tried using `configUtils`, but this utility doesn't work if a default value is defined inside `server/config/index.js` (see https://github.com/TryGhost/Ghost/pull/6861/files#diff-198522c016d444fe14bb9882f7614d2dR237). The `_.merge` is in a wrong order, it merges the default values into the custom values.

Comment (kirrg001) [inline core/test/integration/api/api_schedules_spec.js]: i reset the original value in line `179`.
